File: orders-service/src/alerts/alert-detector.ts
/**
 * Alert Detector for Orders Service
 * 
 * This module detects alert events based on application behavior:
 * - Error bursts (multiple errors in short time)
 * - Repeated failures (same error pattern)
 * - High latency conditions (slow responses)
 * 
 * NO OpenTelemetry, NO Prometheus, NO external observability libraries
 */

import * as fs from 'fs';
import * as path from 'path';

export interface AlertEvent {
  timestamp: string;
  service_name: string;
  alert_name: string;
  alert_type: 'error' | 'latency' | 'availability';
  alert_state: 'fired' | 'resolved';
  alert_duration?: number; // milliseconds, only if resolved
  severity: 'low' | 'medium' | 'high';
  
  // Context fields captured at alert time
  request_count: number;
  error_count: number;
  average_response_time: number;
  process_cpu_usage: number;
  process_memory_usage: number;
}

interface RequestMetrics {
  timestamp: number;
  duration: number;
  isError: boolean;
  errorType?: string;
}

interface ActiveAlert {
  alert_name: string;
  alert_type: 'error' | 'latency' | 'availability';
  fired_at: number;
  severity: 'low' | 'medium' | 'high';
}

export class AlertDetector {
  private serviceName: string;
  private alertDataFile: string;
  private recentRequests: RequestMetrics[] = [];
  private activeAlerts: Map<string, ActiveAlert> = new Map();
  
  // Configuration thresholds
  private readonly ERROR_BURST_THRESHOLD = 5; // errors in window
  private readonly ERROR_BURST_WINDOW = 60000; // 1 minute
  private readonly HIGH_LATENCY_THRESHOLD = 3000; // 3 seconds
  private readonly HIGH_LATENCY_COUNT = 3; // consecutive slow requests
  private readonly AVAILABILITY_ERROR_RATE = 0.5; // 50% error rate
  private readonly METRICS_WINDOW = 300000; // 5 minutes
  
  constructor(serviceName: string) {
    this.serviceName = serviceName;
    
    // Create alert data directory if it doesn't exist
    const alertDir = path.join(process.cwd(), 'logs', 'alert');
    if (!fs.existsSync(alertDir)) {
      fs.mkdirSync(alertDir, { recursive: true });
    }
    
    this.alertDataFile = path.join(alertDir, `${serviceName}-alert-data.ndjson`);
    
    // Start periodic check
    this.startPeriodicCheck();
  }
  
  /**
   * Record a request for alert detection
   */
  public recordRequest(duration: number, isError: boolean, errorType?: string): void {
    const now = Date.now();
    
    this.recentRequests.push({
      timestamp: now,
      duration,
      isError,
      errorType
    });
    
    // Clean old requests outside the metrics window
    this.cleanOldMetrics(now);
    
    // Check for alert conditions immediately
    this.checkAlertConditions();
  }
  
  /**
   * Clean metrics older than the window
   */
  private cleanOldMetrics(now: number): void {
    const cutoff = now - this.METRICS_WINDOW;
    this.recentRequests = this.recentRequests.filter(r => r.timestamp > cutoff);
  }
  
  /**
   * Start periodic alert checking (every 30 seconds)
   */
  private startPeriodicCheck(): void {
    setInterval(() => {
      this.checkAlertConditions();
    }, 30000);
  }
  
  /**
   * Check all alert conditions
   */
  private checkAlertConditions(): void {
    const now = Date.now();
    this.cleanOldMetrics(now);
    
    // Check error burst
    this.checkErrorBurst(now);
    
    // Check high latency
    this.checkHighLatency(now);
    
    // Check availability
    this.checkAvailability(now);
  }
  
  /**
   * Check for error burst condition
   */
  private checkErrorBurst(now: number): void {
    const alertName = 'error_burst';
    const burstWindow = now - this.ERROR_BURST_WINDOW;
    const recentErrors = this.recentRequests.filter(
      r => r.isError && r.timestamp > burstWindow
    );
    
    if (recentErrors.length >= this.ERROR_BURST_THRESHOLD) {
      if (!this.activeAlerts.has(alertName)) {
        // Fire alert
        const severity = recentErrors.length >= 10 ? 'high' : 
                        recentErrors.length >= 7 ? 'medium' : 'low';
        
        this.fireAlert(alertName, 'error', severity);
      }
    } else {
      if (this.activeAlerts.has(alertName)) {
        // Resolve alert
        this.resolveAlert(alertName);
      }
    }
  }
  
  /**
   * Check for high latency condition
   */
  private checkHighLatency(now: number): void {
    const alertName = 'high_latency';
    
    // Get last N requests
    const recentNonError = this.recentRequests
      .filter(r => !r.isError)
      .slice(-this.HIGH_LATENCY_COUNT);
    
    if (recentNonError.length >= this.HIGH_LATENCY_COUNT) {
      const allSlow = recentNonError.every(r => r.duration > this.HIGH_LATENCY_THRESHOLD);
      
      if (allSlow) {
        if (!this.activeAlerts.has(alertName)) {
          const avgLatency = recentNonError.reduce((sum, r) => sum + r.duration, 0) / recentNonError.length;
          const severity = avgLatency > 5000 ? 'high' : avgLatency > 4000 ? 'medium' : 'low';
          
          this.fireAlert(alertName, 'latency', severity);
        }
      } else {
        if (this.activeAlerts.has(alertName)) {
          this.resolveAlert(alertName);
        }
      }
    }
  }
  
  /**
   * Check for availability issues (high error rate)
   */
  private checkAvailability(now: number): void {
    const alertName = 'availability_issue';
    
    if (this.recentRequests.length < 10) {
      // Not enough data
      return;
    }
    
    const errorCount = this.recentRequests.filter(r => r.isError).length;
    const errorRate = errorCount / this.recentRequests.length;
    
    if (errorRate >= this.AVAILABILITY_ERROR_RATE) {
      if (!this.activeAlerts.has(alertName)) {
        const severity = errorRate >= 0.8 ? 'high' : errorRate >= 0.65 ? 'medium' : 'low';
        
        this.fireAlert(alertName, 'availability', severity);
      }
    } else {
      if (this.activeAlerts.has(alertName)) {
        this.resolveAlert(alertName);
      }
    }
  }
  
  /**
   * Fire an alert
   */
  private fireAlert(alertName: string, alertType: 'error' | 'latency' | 'availability', severity: 'low' | 'medium' | 'high'): void {
    const now = Date.now();
    
    this.activeAlerts.set(alertName, {
      alert_name: alertName,
      alert_type: alertType,
      fired_at: now,
      severity
    });
    
    const alertEvent: AlertEvent = {
      timestamp: new Date(now).toISOString(),
      service_name: this.serviceName,
      alert_name: alertName,
      alert_type: alertType,
      alert_state: 'fired',
      severity,
      ...this.captureContext()
    };
    
    this.writeAlertEvent(alertEvent);
  }
  
  /**
   * Resolve an alert
   */
  private resolveAlert(alertName: string): void {
    const activeAlert = this.activeAlerts.get(alertName);
    if (!activeAlert) return;
    
    const now = Date.now();
    const duration = now - activeAlert.fired_at;
    
    this.activeAlerts.delete(alertName);
    
    const alertEvent: AlertEvent = {
      timestamp: new Date(now).toISOString(),
      service_name: this.serviceName,
      alert_name: alertName,
      alert_type: activeAlert.alert_type,
      alert_state: 'resolved',
      alert_duration: duration,
      severity: activeAlert.severity,
      ...this.captureContext()
    };
    
    this.writeAlertEvent(alertEvent);
  }
  
  /**
   * Capture context metrics at alert time
   */
  private captureContext(): {
    request_count: number;
    error_count: number;
    average_response_time: number;
    process_cpu_usage: number;
    process_memory_usage: number;
  } {
    const requestCount = this.recentRequests.length;
    const errorCount = this.recentRequests.filter(r => r.isError).length;
    const avgResponseTime = requestCount > 0
      ? this.recentRequests.reduce((sum, r) => sum + r.duration, 0) / requestCount
      : 0;
    
    // Get process metrics
    const memUsage = process.memoryUsage();
    const cpuUsage = process.cpuUsage();
    
    return {
      request_count: requestCount,
      error_count: errorCount,
      average_response_time: Math.round(avgResponseTime),
      process_cpu_usage: cpuUsage.user + cpuUsage.system,
      process_memory_usage: memUsage.heapUsed
    };
  }
  
  /**
   * Write alert event to file (append-only NDJSON)
   */
  private writeAlertEvent(alertEvent: AlertEvent): void {
    const line = JSON.stringify(alertEvent) + '\n';
    
    fs.appendFile(this.alertDataFile, line, (err) => {
      if (err) {
        console.error(`Failed to write alert event: ${err.message}`);
      }
    });
  }
  
  /**
   * Get current alert statistics (for monitoring)
   */
  public getStats(): {
    activeAlerts: number;
    recentRequests: number;
    recentErrors: number;
  } {
    return {
      activeAlerts: this.activeAlerts.size,
      recentRequests: this.recentRequests.length,
      recentErrors: this.recentRequests.filter(r => r.isError).length
    };
  }
}

File: users-service/src/main/java/com/app/usersservice/collector/AlertDetector.java
package com.app.usersservice.collector;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.mongodb.client.MongoClient;
import jakarta.annotation.PreDestroy;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.lang.management.ManagementFactory;
import java.lang.management.OperatingSystemMXBean;
import java.time.Instant;
import java.util.*;
import java.util.concurrent.*;
import java.util.stream.Collectors;

@Component
public class AlertDetector {
    
    private final String serviceName;
    private final String alertDataFile;
    private final List<RequestMetrics> recentRequests;
    private final Map<String, ActiveAlert> activeAlerts;
    private final ObjectMapper objectMapper;
    private final ScheduledExecutorService scheduler;
    private final List<AuthFailure> authFailures;
    private final List<TrafficRate> trafficRateHistory;
    private double baselineTrafficRate = 0.0;
    private boolean baselineCalculated = false;
    
    @Autowired(required = false)
    private MongoClient mongoClient;
    
    // Configuration thresholds - Original
    private static final int ERROR_BURST_THRESHOLD = 5;
    private static final long ERROR_BURST_WINDOW = 60000; // 1 minute
    private static final long HIGH_LATENCY_THRESHOLD = 3000; // 3 seconds
    private static final int HIGH_LATENCY_COUNT = 3;
    private static final double AVAILABILITY_ERROR_RATE = 0.5; // 50%
    private static final long METRICS_WINDOW = 300000; // 5 minutes
    
    // Configuration thresholds - New
    private static final double MEMORY_THRESHOLD_PERCENT = 85.0; // 85%
    private static final double CPU_THRESHOLD_PERCENT = 80.0; // 80%
    private static final double TRAFFIC_SPIKE_MULTIPLIER = 3.0; // 3x
    private static final double TRAFFIC_DROP_MULTIPLIER = 0.3; // 30%
    private static final long TRAFFIC_BASELINE_WINDOW = 600000; // 10 minutes
    private static final int AUTH_FAILURE_THRESHOLD = 10;
    private static final long AUTH_FAILURE_WINDOW = 300000; // 5 minutes
    
    public AlertDetector() {
        this.serviceName = "users-service";
        this.recentRequests = new CopyOnWriteArrayList<>();
        this.activeAlerts = new ConcurrentHashMap<>();
        this.authFailures = new CopyOnWriteArrayList<>();
        this.trafficRateHistory = new CopyOnWriteArrayList<>();
        this.objectMapper = new ObjectMapper();
        
        // Create alert data directory
        File alertDir = new File("alerts");
        if (!alertDir.exists()) {
            alertDir.mkdirs();
        }
        
        this.alertDataFile = "alerts/" + serviceName + "-alert-data.ndjson";
        
        // Start periodic check (every 30 seconds)
        this.scheduler = Executors.newSingleThreadScheduledExecutor();
        this.scheduler.scheduleAtFixedRate(
            this::checkAlertConditions,
            30,
            30,
            TimeUnit.SECONDS
        );
        
        System.out.println("[Alert Detector] Initialized for " + serviceName);
    }
    
    @PreDestroy
    public void cleanup() {
        if (scheduler != null && !scheduler.isShutdown()) {
            scheduler.shutdown();
            try {
                if (!scheduler.awaitTermination(5, TimeUnit.SECONDS)) {
                    scheduler.shutdownNow();
                }
            } catch (InterruptedException e) {
                scheduler.shutdownNow();
                Thread.currentThread().interrupt();
            }
        }
    }
    
    public void recordRequest(long duration, boolean isError, String errorType) {
        long now = System.currentTimeMillis();
        RequestMetrics metrics = new RequestMetrics(now, duration, isError, errorType);
        recentRequests.add(metrics);
        cleanOldMetrics(now);
        checkAlertConditions();
    }
    
    public void recordAuthFailure(String failureType) {
        authFailures.add(new AuthFailure(System.currentTimeMillis(), failureType));
    }
    
    private void cleanOldMetrics(long now) {
        long cutoff = now - METRICS_WINDOW;
        recentRequests.removeIf(r -> r.timestamp < cutoff);
    }
    
    private void checkAlertConditions() {
        long now = System.currentTimeMillis();
        cleanOldMetrics(now);
        
        // Original alerts
        checkErrorBurst(now);
        checkHighLatency(now);
        checkAvailability(now);
        
        // New resource alerts
        checkMemoryUsage(now);
        checkCPUUsage(now);
        
        // Traffic anomaly alerts
        checkTrafficAnomaly(now);
        
        // Security alerts
        checkAuthFailures(now);
        
        // Database alerts
        checkDatabaseConnection(now);
    }
    
    private void checkErrorBurst(long now) {
        String alertName = "error_burst";
        long burstWindow = now - ERROR_BURST_WINDOW;
        
        long recentErrorCount = recentRequests.stream()
            .filter(r -> r.isError && r.timestamp > burstWindow)
            .count();
        
        if (recentErrorCount >= ERROR_BURST_THRESHOLD) {
            if (!activeAlerts.containsKey(alertName)) {
                String severity = recentErrorCount >= 10 ? "high" :
                                 recentErrorCount >= 7 ? "medium" : "low";
                fireAlert(alertName, "error", severity);
            }
        } else {
            if (activeAlerts.containsKey(alertName)) {
                resolveAlert(alertName);
            }
        }
    }
    
    private void checkHighLatency(long now) {
        String alertName = "high_latency";
        
        List<RequestMetrics> recentNonError = recentRequests.stream()
            .filter(r -> !r.isError)
            .collect(Collectors.toList());
        
        if (recentNonError.size() < HIGH_LATENCY_COUNT) {
            return;
        }
        
        List<RequestMetrics> lastN = recentNonError.subList(
            Math.max(0, recentNonError.size() - HIGH_LATENCY_COUNT),
            recentNonError.size()
        );
        
        boolean allSlow = lastN.stream()
            .allMatch(r -> r.duration > HIGH_LATENCY_THRESHOLD);
        
        if (allSlow) {
            if (!activeAlerts.containsKey(alertName)) {
                double avgLatency = lastN.stream()
                    .mapToLong(r -> r.duration)
                    .average()
                    .orElse(0);
                
                String severity = avgLatency > 5000 ? "high" :
                                 avgLatency > 4000 ? "medium" : "low";
                fireAlert(alertName, "latency", severity);
            }
        } else {
            if (activeAlerts.containsKey(alertName)) {
                resolveAlert(alertName);
            }
        }
    }
    
    private void checkAvailability(long now) {
        String alertName = "availability_issue";
        
        if (recentRequests.size() < 10) {
            return;
        }
        
        long errorCount = recentRequests.stream()
            .filter(r -> r.isError)
            .count();
        
        double errorRate = (double) errorCount / recentRequests.size();
        
        if (errorRate >= AVAILABILITY_ERROR_RATE) {
            if (!activeAlerts.containsKey(alertName)) {
                String severity = errorRate >= 0.8 ? "high" :
                                 errorRate >= 0.65 ? "medium" : "low";
                fireAlert(alertName, "availability", severity);
            }
        } else {
            if (activeAlerts.containsKey(alertName)) {
                resolveAlert(alertName);
            }
        }
    }
    
    private void checkMemoryUsage(long now) {
        String alertName = "high_memory_usage";
        Runtime runtime = Runtime.getRuntime();
        long maxMemory = runtime.maxMemory();
        long usedMemory = runtime.totalMemory() - runtime.freeMemory();
        double usagePercent = (double) usedMemory / maxMemory * 100;
        
        if (usagePercent >= MEMORY_THRESHOLD_PERCENT) {
            if (!activeAlerts.containsKey(alertName)) {
                String severity = usagePercent >= 95 ? "critical" :
                                 usagePercent >= 90 ? "high" : "medium";
                fireAlert(alertName, "resource", severity);
            }
        } else if (usagePercent < MEMORY_THRESHOLD_PERCENT - 5) {
            if (activeAlerts.containsKey(alertName)) {
                resolveAlert(alertName);
            }
        }
    }
    
    private void checkCPUUsage(long now) {
        String alertName = "high_cpu_usage";
        
        double cpuUsage = -1.0;
        OperatingSystemMXBean osBean = ManagementFactory.getOperatingSystemMXBean();
        if (osBean instanceof com.sun.management.OperatingSystemMXBean) {
            com.sun.management.OperatingSystemMXBean sunOsBean = 
                (com.sun.management.OperatingSystemMXBean) osBean;
            double processCpuLoad = sunOsBean.getProcessCpuLoad();
            if (processCpuLoad >= 0) {
                cpuUsage = processCpuLoad * 100;
            }
        }
        
        if (cpuUsage >= CPU_THRESHOLD_PERCENT) {
            if (!activeAlerts.containsKey(alertName)) {
                String severity = cpuUsage >= 95 ? "critical" :
                                 cpuUsage >= 90 ? "high" : "medium";
                fireAlert(alertName, "resource", severity);
            }
        } else if (cpuUsage >= 0 && cpuUsage < CPU_THRESHOLD_PERCENT - 10) {
            if (activeAlerts.containsKey(alertName)) {
                resolveAlert(alertName);
            }
        }
    }
    
    private void checkTrafficAnomaly(long now) {
        long oneMinuteAgo = now - 60000;
        long recentTrafficCount = recentRequests.stream()
            .filter(r -> r.timestamp > oneMinuteAgo)
            .count();
        double currentRate = recentTrafficCount / 60.0;
        
        trafficRateHistory.add(new TrafficRate(now, recentTrafficCount));
        
        // Keep only last 10 minutes
        trafficRateHistory.removeIf(t -> t.timestamp <= now - TRAFFIC_BASELINE_WINDOW);
        
        // Calculate baseline
        if (!baselineCalculated && trafficRateHistory.size() >= 5) {
            baselineTrafficRate = trafficRateHistory.stream()
                .mapToDouble(t -> t.count / 60.0)
                .average()
                .orElse(0);
            baselineCalculated = true;
        }
        
        if (baselineCalculated && baselineTrafficRate > 0) {
            // Check for traffic spike
            if (currentRate >= baselineTrafficRate * TRAFFIC_SPIKE_MULTIPLIER) {
                String alertName = "traffic_spike";
                if (!activeAlerts.containsKey(alertName)) {
                    double multiplier = currentRate / baselineTrafficRate;
                    String severity = multiplier >= 5 ? "high" :
                                     multiplier >= 4 ? "medium" : "low";
                    fireAlert(alertName, "traffic", severity);
                }
            } else if (currentRate < baselineTrafficRate * (TRAFFIC_SPIKE_MULTIPLIER - 0.5)) {
                String alertName = "traffic_spike";
                if (activeAlerts.containsKey(alertName)) {
                    resolveAlert(alertName);
                }
            }
            
            // Check for traffic drop
            if (currentRate <= baselineTrafficRate * TRAFFIC_DROP_MULTIPLIER) {
                String alertName = "traffic_drop";
                if (!activeAlerts.containsKey(alertName)) {
                    double dropPercent = (1 - currentRate / baselineTrafficRate) * 100;
                    String severity = dropPercent >= 90 ? "critical" :
                                     dropPercent >= 80 ? "high" :
                                     dropPercent >= 70 ? "medium" : "low";
                    fireAlert(alertName, "traffic", severity);
                }
            } else if (currentRate > baselineTrafficRate * (TRAFFIC_DROP_MULTIPLIER + 0.2)) {
                String alertName = "traffic_drop";
                if (activeAlerts.containsKey(alertName)) {
                    resolveAlert(alertName);
                }
            }
        }
    }
    
    private void checkAuthFailures(long now) {
        String alertName = "auth_failure_spike";
        
        // Clean old auth failures
        authFailures.removeIf(f -> f.timestamp <= now - AUTH_FAILURE_WINDOW);
        
        if (authFailures.size() >= AUTH_FAILURE_THRESHOLD) {
            if (!activeAlerts.containsKey(alertName)) {
                String severity = authFailures.size() >= 50 ? "critical" :
                                 authFailures.size() >= 30 ? "high" :
                                 authFailures.size() >= 20 ? "medium" : "low";
                fireAlert(alertName, "security", severity);
            }
        } else if (authFailures.size() < AUTH_FAILURE_THRESHOLD - 3) {
            if (activeAlerts.containsKey(alertName)) {
                resolveAlert(alertName);
            }
        }
    }
    
    private void checkDatabaseConnection(long now) {
        String alertName = "database_connection_issue";
        
        boolean isConnected = false;
        if (mongoClient != null) {
            try {
                // Try to ping the database
                mongoClient.listDatabaseNames().first();
                isConnected = true;
            } catch (Exception e) {
                isConnected = false;
            }
        }
        
        if (!isConnected) {
            if (!activeAlerts.containsKey(alertName)) {
                fireAlert(alertName, "resource", "critical");
            }
        } else {
            if (activeAlerts.containsKey(alertName)) {
                resolveAlert(alertName);
            }
        }
    }
    
    private void fireAlert(String alertName, String alertType, String severity) {
        long now = System.currentTimeMillis();
        
        ActiveAlert activeAlert = new ActiveAlert(alertName, alertType, now, severity);
        activeAlerts.put(alertName, activeAlert);
        
        AlertEvent alertEvent = new AlertEvent();
        alertEvent.setTimestamp(Instant.now().toString());
        alertEvent.setServiceName(serviceName);
        alertEvent.setAlertName(alertName);
        alertEvent.setAlertType(alertType);
        alertEvent.setAlertState("fired");
        alertEvent.setSeverity(severity);
        
        // Capture context
        ContextMetrics context = captureContext();
        alertEvent.setRequestCount(context.requestCount);
        alertEvent.setErrorCount(context.errorCount);
        alertEvent.setAverageResponseTime(context.averageResponseTime);
        alertEvent.setProcessCpuUsage(context.processCpuUsage);
        alertEvent.setProcessMemoryUsage(context.processMemoryUsage);
        alertEvent.setTrafficRate(context.trafficRate);
        
        writeAlertEvent(alertEvent);
    }
    
    private void resolveAlert(String alertName) {
        ActiveAlert activeAlert = activeAlerts.get(alertName);
        if (activeAlert == null) return;
        
        long now = System.currentTimeMillis();
        long duration = now - activeAlert.firedAt;
        
        activeAlerts.remove(alertName);
        
        AlertEvent alertEvent = new AlertEvent();
        alertEvent.setTimestamp(Instant.now().toString());
        alertEvent.setServiceName(serviceName);
        alertEvent.setAlertName(alertName);
        alertEvent.setAlertType(activeAlert.alertType);
        alertEvent.setAlertState("resolved");
        alertEvent.setAlertDuration(duration);
        alertEvent.setSeverity(activeAlert.severity);
        
        // Capture context
        ContextMetrics context = captureContext();
        alertEvent.setRequestCount(context.requestCount);
        alertEvent.setErrorCount(context.errorCount);
        alertEvent.setAverageResponseTime(context.averageResponseTime);
        alertEvent.setProcessCpuUsage(context.processCpuUsage);
        alertEvent.setProcessMemoryUsage(context.processMemoryUsage);
        alertEvent.setTrafficRate(context.trafficRate);
        
        writeAlertEvent(alertEvent);
    }
    
    private ContextMetrics captureContext() {
        int requestCount = recentRequests.size();
        int errorCount = (int) recentRequests.stream()
            .filter(r -> r.isError)
            .count();
        
        long avgResponseTime = requestCount > 0
            ? (long) recentRequests.stream()
                .mapToLong(r -> r.duration)
                .average()
                .orElse(0)
            : 0;
        
        // Get process metrics
        Runtime runtime = Runtime.getRuntime();
        long memoryUsage = runtime.totalMemory() - runtime.freeMemory();
        
        double cpuUsage = -1.0;
        OperatingSystemMXBean osBean = ManagementFactory.getOperatingSystemMXBean();
        if (osBean instanceof com.sun.management.OperatingSystemMXBean) {
            com.sun.management.OperatingSystemMXBean sunOsBean = 
                (com.sun.management.OperatingSystemMXBean) osBean;
            double processCpuLoad = sunOsBean.getProcessCpuLoad();
            if (processCpuLoad >= 0) {
                cpuUsage = processCpuLoad * 100;
            }
        }
        
        if (cpuUsage < 0) {
            double loadAvg = osBean.getSystemLoadAverage();
            cpuUsage = loadAvg >= 0 ? loadAvg : -1.0;
        }
        
        // Calculate traffic rate
        long now = System.currentTimeMillis();
        long oneMinuteAgo = now - 60000;
        long recentTrafficCount = recentRequests.stream()
            .filter(r -> r.timestamp > oneMinuteAgo)
            .count();
        double trafficRate = recentTrafficCount / 60.0;
        
        return new ContextMetrics(requestCount, errorCount, avgResponseTime, 
                                 cpuUsage, memoryUsage, trafficRate);
    }
    
    private synchronized void writeAlertEvent(AlertEvent alertEvent) {
        try {
            String json = objectMapper.writeValueAsString(alertEvent);
            
            try (FileWriter writer = new FileWriter(alertDataFile, true)) {
                writer.write(json);
                writer.write(System.lineSeparator());
            }
        } catch (IOException e) {
            System.err.println("Failed to write alert event: " + e.getMessage());
        }
    }
    
    public Map<String, Object> getStats() {
        Map<String, Object> stats = new HashMap<>();
        stats.put("activeAlerts", activeAlerts.size());
        stats.put("recentRequests", recentRequests.size());
        stats.put("recentErrors", recentRequests.stream().filter(r -> r.isError).count());
        return stats;
    }
    
    // Inner classes
    private static class RequestMetrics {
        long timestamp;
        long duration;
        boolean isError;
        String errorType;
        
        RequestMetrics(long timestamp, long duration, boolean isError, String errorType) {
            this.timestamp = timestamp;
            this.duration = duration;
            this.isError = isError;
            this.errorType = errorType;
        }
    }
    
    private static class ActiveAlert {
        String alertName;
        String alertType;
        long firedAt;
        String severity;
        
        ActiveAlert(String alertName, String alertType, long firedAt, String severity) {
            this.alertName = alertName;
            this.alertType = alertType;
            this.firedAt = firedAt;
            this.severity = severity;
        }
    }
    
    private static class ContextMetrics {
        int requestCount;
        int errorCount;
        long averageResponseTime;
        double processCpuUsage;
        long processMemoryUsage;
        double trafficRate;
        
        ContextMetrics(int requestCount, int errorCount, long averageResponseTime,
                      double processCpuUsage, long processMemoryUsage, double trafficRate) {
            this.requestCount = requestCount;
            this.errorCount = errorCount;
            this.averageResponseTime = averageResponseTime;
            this.processCpuUsage = processCpuUsage;
            this.processMemoryUsage = processMemoryUsage;
            this.trafficRate = trafficRate;
        }
    }
    
    private static class AuthFailure {
        long timestamp;
        String type;
        
        AuthFailure(long timestamp, String type) {
            this.timestamp = timestamp;
            this.type = type;
        }
    }
    
    private static class TrafficRate {
        long timestamp;
        long count;
        
        TrafficRate(long timestamp, long count) {
            this.timestamp = timestamp;
            this.count = count;
        }
    }
}
